src/main/java/org/sagebionetworks/bridge/models/schedules2/timelines/Scheduler.java

```diff
@@ -0,0 +1,159 @@
+        if (expiration == null) {
+            return studyLengthInDays;
+        }
+        int expInMinutes = expiration.toStandardMinutes().getMinutes();
+        int minutesInDay = (startTime.getHourOfDay() * 60) + startTime.getMinuteOfHour();
+        
+        int endDay = startDay + ((minutesInDay + expInMinutes)/60/24);
+        // If the time of day is 00:00, then don't return the end day because no part of it can be used
+        // by the participant, and there's no reason for the client to pull the scheduled session for 
+        // this day.
+        if ((minutesInDay + expInMinutes) % 1440 == 0) {
+            endDay -= 1;
+        }
+        return endDay;
+    }
+
+    void scheduleTimeWindowSequence(Timeline.Builder builder, Schedule2 schedule, Session session, TimeWindow window) {
+        int studyLengthInDays = schedule.getDuration().toStandardDays().getDays(); // can be in days or weeks
+        
+        int delayInDays = (session.getDelay() == null) ? 0 : session.getDelay().toStandardDays().getDays();
+        int intervalInDays = (session.getInterval() == null) ? 0 : session.getInterval().toStandardDays().getDays();
+        int occurrenceMax = (session.getOccurrences() == null) ? Integer.MAX_VALUE : session.getOccurrences();
+        
+        int startDay = delayInDays;
+        int endDay = delayInDays;
+        int occurrenceCount = 0;
+
+        Multiset<String> ids = HashMultiset.create();
+        do {
+            endDay = startDay;
+            
+            LocalTime startTime = window.getStartTime();
+            Period delay = session.getDelay();
+            Period expiration = window.getExpiration();
+            
+            endDay = calculateEndDay(studyLengthInDays, startTime, startDay, expiration);
+            // If it extends beyond the end of the study, it is not included in timeline
+            if (endDay > studyLengthInDays) {
+                break;
+            }
+            
+            // This session will be used, so we can add it
+            builder.withSessionInfo(SessionInfo.create(session));
+            
+            ScheduledSession.Builder scheduledSession = new ScheduledSession.Builder();
+            scheduledSession.withRefGuid(session.getGuid());
+            scheduledSession.withStartDay(startDay);
+            scheduledSession.withEndDay(endDay);
+            scheduledSession.withStartTime(window.getStartTime());
+            scheduledSession.withExpiration(window.getExpiration());
+            scheduledSession.withPersistent(window.isPersistent());
+            // If we have a delay under one day, and it's the first day, then skip the startTime 
+            // and delay for the indicated time before showing.
+            if (startDay == 0 && delay != null && delay.toStandardDays().getDays() == 0) {
+                scheduledSession.withDelayTime(delay);
+                scheduledSession.withStartTime(null);
+            }
```

Comment:
This is ridiculously complex. I can't even describe it so what should we do here?

Reply:
I left in the delayTime for this particular day, but also left in startTime, so it's now just an additional piece of information that the client an ignore or interpret as it will.